### DISCUSSION TITLE
How to synchronise dark/light mode between docusaurus and MaterialUI?

### DISCUSSION THREAD

**Discussion by dcs3spp** [facebook/docusaurus#8694] [Q&A] [answered]
Hi,

I am new Docusaurus user with newbie React knowledge. I am trying to
synchronise the docusaurus theme with MaterialUI.

My approach so far has been to swizzle docusaurus `ColorModeToggle` via
wrapping. From the wrapped `ColorodeToggle` I retrieve a function stored in a
React context to toggle the light/dark theme in MaterialUI. Within the swizzled
`Root` I use the react context provider which , in turn, wraps a materialUI
`ThemeProvider` For further details I have included the code below.

However when I browse my site I get the following error:

```
Maximum update depth exceeded. This can happen when a component repeatedly calls
setState inside componentWillUpdate or componentDidUpdate. React limits the
number of nested updates to prevent infinite loops.
```

Has anyone else managed to synchronise docusaurus light/dark theme with
MaterialUI?

**The swizzled ColorModeToggle**
```javascript
import React from "react";
import ColorModeToggle from "@theme-original/ColorModeToggle";
import { useToggleTheme } from "@site/src/components/MuiTheme";

export default function ColorModeToggleWrapper(props) {
  console.log("<ColorModeToggleWrapper> properties = " + JSON.stringify(props));

  // "value" holds docusaurus color theme. Either "light" or "dark"
  const { value } = props;
  const muiToggle = useToggleTheme();

  console.log("Docusaurus theme = " + value);
  console.log("MUI theme dark = " + muiToggle());

  return (
    <>
      <ColorModeToggle {...props} />
    </>
  );
}
```

**The React Context**
```javascript
import React, { useContext } from "react";
import { createTheme, ThemeProvider } from "@mui/material/styles";

const CustomThemeContext = React.createContext({ toggleTheme: () => {} });

const darkTheme = createTheme({
  components: {
    MuiListItemText: {
      styleOverrides: {
        primary: {
          color: "orange",
        },
        secondary: {
          color: "purple",
        },
      },
    },
  },
  palette: {
    mode: "dark",
    primary: {
      main: "hsl(8,71%,28%)" /* burgundy mapped to link */,
    },
    secondary: {
      main: "hsl(61,78%,26%)" /* brown */,
    },
  },
});

const lightTheme = createTheme({
  components: {
    MuiListItemText: {
      styleOverrides: {
        primary: {
          color: "aqua",
        },
        secondary: {
          color: "grey",
        },
      },
    },
  },
  palette: {
    mode: "light",
    primary: {
      main: "hsl(8,10%,18%)" /* burgundy mapped to link */,
    },
    secondary: {
      main: "hsl(61,18%,26%)" /* brown */,
    },
  },
});

export function CustomThemeProvider({ children }) {
  const [dark, setDark] = React.useState(false);

  function toggleTheme() {
    console.log("toggleTheme :: from dark = " + dark);
    if (dark === true) {
      setDark(false);
    } else {
      setDark(true);
    }
  }

  const theme = React.useMemo(() => {
    if (dark === true) {
      return createTheme(darkTheme);
    }
    return createTheme(lightTheme);
  }, [dark]);

  return (
    <CustomThemeContext.Provider value={toggleTheme}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </CustomThemeContext.Provider>
  );
}

export function useToggleTheme() {
  const context = useContext(CustomThemeContext);
  if (context === undefined) {
    throw new Error(
      "useCustomThemeContext must be used within an CustomThemeProvider"
    );
  }
  return context;
}
```

**The Root component**
```javascript
import React from "react";
import { CustomThemeProvider } from "@site/src/components/MuiTheme";
import App from "@site/src/components/App";

export default function Root({ children }) {
  return (
    <>
      <CustomThemeProvider>
        <App children={children}></App>
      </CustomThemeProvider>
    </>
  );
}
```

**The App component**
```javascript
import React from "react";

export default function App(props) {
  return <React.Fragment>{props.children}</React.Fragment>;
}
```

**Comment by dcs3spp** [comment, marked as answer]
Solved why the error message was happening. I was calling the `toggleTheme` function directly during rendering. Further details are explained [here](https://typeofnan.dev/fix-the-maximum-update-depth-exceeded-error-in-react/)

I updated the `ColorModeToggleWrapper` to be as listed below. This uses `useEffect` to monitor change of the docusaurus value property from the wrapper component properties. When the value changes then `useEffect` toggles the material UI theme via using a reference to the function stored in the react context.

```javascript
import React, { useEffect } from "react";
import ColorModeToggle from "@theme-original/ColorModeToggle";
import { useToggleTheme } from "@site/src/components/MuiTheme";

export default function ColorModeToggleWrapper(props) {
  // extract the docusaurus theme from the component properties
  const { value } = props;

  // get the toggleTheme function from the context
  const toggleTheme = useToggleTheme();

  // whenever the theme changes in docusaurus trigger the change
  // in MaterialUI by calling the callback function stored in the react
  // context
  useEffect(() => {
    console.log("Docusaurus theme changed to = " + value);
    console.log("Synching theme change with MaterialUI");

    toggleTheme();
  }, [value]);

  return (
    <>
      <ColorModeToggle {...props} />
    </>
  );
}
```

This solved the error detailed in the question. Now I just need to ensure that the initial starting state of the MaterialUI theme matches the initial state of the docusaurus theme......

**Reply by slorber** [reply]
> Now I just need to ensure that the initial starting state of the MaterialUI theme matches the initial state of the docusaurus theme......

Docusaurus does not support CSS-in-JS libs well due to the inability to extract the critical CSS those libs produce. This can lead to a flash of unstyled things on site load. More info here: https://github.com/facebook/docusaurus/issues/3236

I wouldn't recommend using MUI with Docusaurus for things affecting your site layout. If you want to document a MUI design system in Docusaurus, you could delay the rendering of those components so that they are not server-rendered, and also wait for the theme of both to be in sync.

Note if you want to know the effective theme Docusauurs is using asap, you can try to get this info from `<html data-theme="">` attribute.

**Reply by dcs3spp** [reply]
Hi @slorber thanks for this, appreciated :) Thanks, I will update my custom MUI theme provider to inspect the data-theme attribute from within the html tag.

In the personal site I am building I am using MUI for the landing page in Docusaurus. I got the above code working on a local branch and seems ok locally on page load. However, have not yet published this live. 

The only thing I want to synchronise between Docusaurus and MUI would be the primary and secondary theme colours from ifm->MUI.  

So far, I managed to access the colours within client side component via `window.getComputedStyle(document.documentElement)`. However, as soon as I extract this into a utility function inside it's own module, the Docusaurus server build breaks with error:

> Reference error: window is not defined

I am guessing this is because the module is processed server side? I posted this [question](https://stackoverflow.com/q/75545965/12152982) on Stack Overflow.

> Docusaurus does not support CSS-in-JS libs well due to the inability to extract the critical CSS those libs produce. This can lead to a flash of unstyled things on site load. More info here: https://github.com/facebook/docusaurus/issues/3236

> I wouldn't recommend using MUI with Docusaurus for things affecting your site layout. If you want to document a MUI design system in Docusaurus, you could delay the rendering of those components so that they are not server-rendered, and also wait for the theme of both to be in sync.

Before I proceed further with my implementation, will synching the primary/secondary colours from within the swizzled `ColorModeToggle` cause any flash of unstyled things on site load?

If so, how would I delay the rendering of a component so that it is not server rendered?

**Reply by slorber** [reply]
> In the personal site I am building I am using MUI for the landing page in Docusaurus. I got the above code working on a local branch and seems ok locally on page load. However, have not yet published this live.

Make sure to run `docusaurus build` locally to ensure what you do works fine in production more. You might be disappointed, I wouldn't recommend using MUI for a landing page.

> Before I proceed further with my implementation, will synching the primary/secondary colours from within the swizzled ColorModeToggle cause any flash of unstyled things on site load?

Yes, very likely you'll see your landing page with missing styles, and the styles will load afterwards, leading to a bad UX.

> If so, how would I delay the rendering of a component so that it is not server rendered?

You can create your own wrapper component, or use our `<BrowserOnly>` component. 

But all this is a bad idea for a landing page: you want your content to be server-rendered appropriately for many reasons, including UX, SEO...

---

> So far, I managed to access the colours within client side component via window.getComputedStyle(document.documentElement). However, as soon as I extract this into a utility function inside it's own module, the Docusaurus server build breaks with error:

Your app is loaded in both nodejs and the browser, you must be careful when using apis like `document` or `window` and program defensively: these globals won't be available in nodejs so you shouldn't access them without having a condition such as `typeof window !== 'undefined'` or something similar.

**Reply by dcs3spp** [reply]
Many thanks @slorber for your advice and help. Much appreciated and insightful. Ok, understood, I will take the points on board and will reconsider the use of MUI 👍
